crypto/stark-verifier-ethereum/contracts/primefield.sol

```diff
@@ -176,4 +195,109 @@ library PrimeField {
+    //     inv_prod = inv_prod.fmul(batch_in[i]);
+    //     assembly {
+    //         i := sub(i, 1)
+    //     }
+    // }
+    // batch_out[0] = inv_prod;
+
+    uint256 constant MODULUS_SUB_2 = 0x0800000000000010ffffffffffffffffffffffffffffffffffffffffffffffff;
+
+    // This is a pure assembly optiomized version of a batch inversion
+    // If the batch inversion input data array contains a zero, the batch
+    // inversion will fail.
+    // TODO - Inplace version/ version without output array?
+    function batch_invert(uint256[] memory input_data, uint256[] memory output_data) internal returns(uint256 result) {
+        require(input_data.length == output_data.length);
+
+        assembly {
+            // Uses the fact that data^p = data => data^(p-2) * data = 1
+            // to calculate the multiplicative inverse in the field
+            function invert(data) -> invert_result {
+                let p := mload(0x40)
+                mstore(p, 0x20) // Length of Base
+                mstore(add(p, 0x20), 0x20) // Length of Exponent
+                mstore(add(p, 0x40), 0x20) // Length of Modulus
+                mstore(add(p, 0x60), data) // Base
+                mstore(add(p, 0x80), MODULUS_SUB_2) // Exponent
+                mstore(add(p, 0xa0), MODULUS) // Modulus
+                // call modexp precompile
+                if iszero(call(not(0), 0x05, 0, p, 0xc0, p, 0x20)) {
+                    revert(0, 0)
+                }
+                invert_result := mload(p)
+            }
+
+            let carried := 1
+
+            // This local copy of pointers to data
+            // will be manipulated instead of the real thing
+            let in_pointer := add(input_data, 32)
+            // Note - we don't keep a copy of the output pointer
+            // intead we keep the diffrence between the memory
+            // arrays and use that to adjust the local pointer.
+            // This works no matter memory layout because of the
+            // modularity of evm additon
+            // TODO - does this dif method actually save anything?
+            let out_dif := sub(output_data, input_data)
+
+            // The end bound of the following loop is when it's
+            // 32*len past the data pointer
+            let final_pointer := add(in_pointer, mul(mload(input_data), 32))
+
+            // We interate on the pointer by moving forward
+            // a word at a time and then checking we aren't
+            // beyond the final pointer.
+            for {} lt(in_pointer, final_pointer) {in_pointer := add(in_pointer, 32)} {
+                // We want to get the product of all of the previous
+                // elements into each slot of output data
+                carried := mulmod(carried, mload(in_pointer), MODULUS)
+                // Using the outdif we store into the output array
+                mstore(add(out_dif, in_pointer), carried)
+            }
```

Comment:
Probably not worth it here, but further gas reductions are possible by unrolling this loop. We do this in Horner. Here's it's probably not worth the additional complexity.

Reply:
The batch invert is still taking around 600k of the gas for the whole call, so I might add the unrolling as in horner.